lib/rlib.js

```diff
@@ -68,185 +84,237 @@ module.exports = (function() {
-            // NOTE: toString('binary') is not binary, it converts from latin-1
-            if (readBinary) databuf = appendBuffer(databuf, chunk);
-            else data += chunk;
-            //else chunks.push(chunk);
+    };
+
+
+    function buildReadBody(options) {
+        options = options || {};
+
+        // Default maxBodySize to 100kb if not passed in.
+        // If it is not a number send it to bytes for parsing.
+        if (typeof options.maxBodySize !== "number") {
+            options.maxBodySize = bytes.parse(options.maxBodySize || "100kb");
         }
-        function readloop() {
-            if (!req._bodyEof) {
-                var chunk = req.read();
-                if (chunk) {
-                    //gatherChunk(chunk);
-                    if (readBinary) gatherChunk(chunk);
-                    else data += chunk;
-                }
-                if (!readImmediate) {
-                    // 35% higher peak throughput with setTimeout (2800/s vs 2050)
-                    // 3x higher throughput per connection with setImmediate (1700/s vs 550)
-                    // +10% peak throughput when using qtimers (3000/s, 2300/s)
-                    setTimeout(readloop, 1);
+
+        return function readBody( req, res, next ) {
+            // body is read only once ever, and _bodyEof is the mutex
+            if (req._bodyEof !== undefined) return next ? next() : null;
+            req._bodyEof = false;
+
+            var readErr;
+            var bytesReceived = 0;
+            var data = "";
+            var databuf;
+            //var chunks = new Array();
+            var readBinary = req.restiq._opts.readBinary;
+            var readImmediate = req.restiq._opts.readImmediate;
+
+            // have the system deal with not splitting multi-byte chars
+            if (!readBinary) req.setEncoding('utf8');
+
+            // default to on('data'), not the fastest but more versatile
+            // 2 is 15% less max throughput than 0 and 10% slower than 1,
+            // but scales down better than 0 and doesnt affect gc like 1
+            if (readImmediate === undefined) readImmediate = 2;
+
+            function appendBuffer( databuf, chunk ) {
+                return databuf ? databuf.concat([databuf, chunk]) : chunk;
+                // TODO: find an efficient way of combining contents,
+                // eg preallocate space and copy data into it.
+            }
+            function gatherChunk(chunk) {
+                // NOTE: it is faster to concat strings than to push the buffers
+                // NOTE: toString('binary') is not binary, it converts from latin-1
+                bytesReceived += chunk.length;
+
+                if (bytesReceived > options.maxBodySize) {
+                    req._bodyEof = true;
+                    readErr = new Restiq.errors.ErrorBadRequest('Error reading body, max request body size exceeded.');
```

Comment:
Good catch. If I have some time tomorrow maybe I will open another pull request and help you get that in. Looks like an easy enough change.